MetaMask/metamask-extension
# Split SRP input by word

## Explanation

Split the secret recovery phrase input field into one field per word. Each word uses a `type="password"` field. There is a "show/hide toggle" next to each word allowing you to reveal it, but only one word can be revealed at a time.

This is a new design aimed at making our SRP input component more secure. This design makes it impossible to accidentally reveal your entire SRP to anybody nearby. It also acts as additional protection against the complete SRP being stored in a browser autocomplete database (though to be clear, we could find no evidence this was ever occurring).

## More information

Figma design: https://www.figma.com/file/8z4Csc3wlOIkUz2OP1eTZW/Secure-UX-Improvements?node-id=1241%3A12046

There have been some changes from the original design (e.g. the addition of the tip about pasting). Also that design was for the "Onboarding V2" flow, so it was not followed exactly. For now this PR targets the old onboarding flow. I will add this new component to the new onboarding flow in the next PR after this.

## Screenshots/Screencaps

### Before

![old](https://user-images.githubusercontent.com/2459287/158828560-ad318679-1a00-48ce-a046-abe0016e23db.png)

### After

![new](https://user-images.githubusercontent.com/2459287/158828597-ac3368ea-54d2-4a4e-8f61-203d45472431.png)

https://user-images.githubusercontent.com/2459287/158829027-1f21e96f-f100-4d74-98b7-c734fe582074.mp4

## Manual testing steps

The two screens that have changed are the "Import" page of onboarding, and the "Restore wallet" page accessible from the lock screen. The component and how it behaves should be identical in both places.

Suggestions on which behaviors to test:
 * It should allow manually typing in the SRP, or pasting the entire thing, or pasting each word individually.
 * The clipboard should be cleared on paste, *unless* the contents are too long (>24 words). In that case, the clipboard is preserved and the paste event is ignored (i.e. no fields are changed).
 * This should be fully navigable via keyboard, and the show/hide toggle should be toggle-able using the spacebar while focused.
 * All fields should be hidden on paste.
 * Only one field should be revealed at a time. Revealing any field hides the others.
